File: AKW/frontend/src/api/nodes.ts
import type { ArchitectureNode } from '@/types'
import client from './client'

export const nodesApi = {
  list: (versionId: string) =>
    client.get<ArchitectureNode[]>(`/versions/${versionId}/nodes`).then((r) => r.data),
  get: (id: string) => client.get<ArchitectureNode>(`/nodes/${id}`).then((r) => r.data),
  create: (versionId: string, data: {
    title: string
    type: string
    position?: { x: number; y: number }
    metadata_?: Record<string, unknown>
    parent_id?: string | null
    reason?: string
    author?: string
  }) => client.post<ArchitectureNode>(`/versions/${versionId}/nodes`, data).then((r) => r.data),
  update: (id: string, data: {
    title?: string
    type?: string
    position?: { x: number; y: number }
    metadata_?: Record<string, unknown>
    parent_id?: string | null
    reason?: string
    author?: string
  }) => client.patch<ArchitectureNode>(`/nodes/${id}`, data).then((r) => r.data),
  delete: (id: string, reason?: string, author?: string) =>
    client.delete(`/nodes/${id}?reason=${reason ?? ''}&author=${author ?? 'unknown'}`),
}

File: AKW/backend/app/routers/nodes.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_db, get_current_user
from app.schemas.node import NodeCreate, NodeUpdate, NodeResponse
from app.services import node_service

router = APIRouter(tags=["nodes"], dependencies=[Depends(get_current_user)])


@router.get("/versions/{version_id}/nodes", response_model=list[NodeResponse])
async def list_nodes(version_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    return await node_service.get_version_nodes(db, version_id)


@router.post("/versions/{version_id}/nodes", response_model=NodeResponse, status_code=201)
async def create_node(version_id: uuid.UUID, data: NodeCreate, db: AsyncSession = Depends(get_db)):
    return await node_service.create_node(db, version_id, data)


@router.get("/nodes/{node_id}", response_model=NodeResponse)
async def get_node(node_id: uuid.UUID, db: AsyncSession = Depends(get_db)):
    return await node_service.get_node(db, node_id)


@router.patch("/nodes/{node_id}", response_model=NodeResponse)
async def update_node(node_id: uuid.UUID, data: NodeUpdate, db: AsyncSession = Depends(get_db)):
    return await node_service.update_node(db, node_id, data)


@router.delete("/nodes/{node_id}", status_code=204)
async def delete_node(
    node_id: uuid.UUID,
    reason: str = Query(default=""),
    author: str = Query(default="unknown"),
    db: AsyncSession = Depends(get_db),
):
    await node_service.delete_node(db, node_id, reason, author)
